aspects: set can write to multiple paths

MiguelPires opened the pull request:
Per the spec, we should allow writes to match several entries with the same request but different paths so the storage can change while supporting old and new paths. The matching should also work on prefixes so we may write to paths nested in other matching paths and therefore have to take care to write to them in order of parentage to prevent overwriting more nested values.

andrewphelpsj reviewed (approved):
Looks good! Thanks.

pedronis commented on aspects/aspects.go:
```diff
@@ -314,23 +303,36 @@ func (a *Aspect) Set(databag DataBag, request string, value interface{}) error {
+	subkeys := strings.Split(request, ".")
 	for _, accessPatt := range a.accessPatterns {
-		placeholders, _, ok := accessPatt.match(requestSubkeys)
+		placeholders, _, ok := accessPatt.match(subkeys)
```
what happens with the restSuffix or if there are unmatched placeholders?

pedronis commented on aspects/aspects.go:
```diff
@@ -314,23 +303,36 @@ func (a *Aspect) Set(databag DataBag, request string, value interface{}) error {
 	}
 
-	requestSubkeys := strings.Split(request, ".")
+	var paths []string
```
maybe it would be cleared to call this storagePaths?

pedronis commented on aspects/aspects_test.go:
```diff
@@ -1074,3 +1074,80 @@ func (s *aspectSuite) TestBadRequestPaths(c *C) {
+	aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+		"foo": []map[string]string{
+			{"request": "a", "storage": "a", "access": "write"},
+			{"request": "a.b", "storage": "b", "access": "write"},
```
with this request we would try to write .b of value into b of storage? but that doesn't seem to be what happens, as value here is a not map this wouldn't work anyway

pedronis reviewed (changes requested):
I'm confused about the logic ignoring rest suffixes

MiguelPires commented:
@pedronis as we discussed, I reworked this PR to deal with writes with unmatched parts in a consistent way and added a few read-write-read tests so we can see this working. Set calls with unmatched placeholders errors for now.

pedronis commented on aspects/aspects.go:
```diff
@@ -339,10 +391,12 @@ func (a *Aspect) Set(databag DataBag, request string, value interface{}) error {
 
-		return a.bundle.schema.Validate(data)
+		if err := a.bundle.schema.Validate(data); err != nil {
+			return err
```
should we check this error path?

pedronis commented on aspects/aspects.go:
```diff
@@ -845,14 +899,21 @@ func set(subKeys []string, index int, node map[string]json.RawMessage, value int
-		return nil, err
+	err := jsonutil.DecodeWithNumber(bytes.NewReader(rawLevel), &level)
+	if err != nil {
+		var uerr *json.UnmarshalTypeError
+		if !errors.As(err, &uerr) {
+			return nil, err
+		}
+	}
+
+	// stored valued wasn't map but new write expects one so overwrite value
+	if level == nil {
+		level = make(map[string]json.RawMessage)
```
when does this happen? tests don't hit this code atm

MiguelPires replied:
When we overwrite a leaf node to be more nested. I added a small test for it

pedronis commented on aspects/aspects.go:
```diff
@@ -311,26 +329,60 @@ type Aspect struct {
-			return err
+		for _, part := range suffixParts {
+			if len(part) > 1 && part[0] == '{' && part[len(part)-1] == '}' {
+				return badRequestErrorFrom(a, "set", request, "cannot set with unmatched placeholders")
```
do we want a TODO here? ultimately we want to support this

MiguelPires replied:
Sure, I'll add one

MiguelPires commented on aspects/aspects.go:
```diff
@@ -311,26 +329,60 @@ type Aspect struct {
+		}
+
+		if err := databag.Set(match.storagePath, nestedValue); err != nil {
 			return err
```
When we overwrite a leaf node to be more nested. I added a small test for it

andrewphelpsj commented on aspects/aspects.go:
```diff
@@ -300,17 +329,23 @@ type Aspect struct {
 		}
 
+		for _, part := range suffixParts {
+			if len(part) > 1 && part[0] == '{' && part[len(part)-1] == '}' {
```
I guess this code will go away/change eventually, so it might not matter.

MiguelPires replied:
I changed it anyway, thanks

andrewphelpsj commented on aspects/aspects.go:
```diff
@@ -86,6 +86,35 @@ func notFoundErrorFrom(a *Aspect, request, errMsg string) *NotFoundError {
+	return ok
+}
+
+func badRequestErrorFrom(a *Aspect, operation, request, errMsg string, v ...any) *BadRequestError {
```
We use `interface{}` everywhere else, rather than `any`. Although at some point, swapping might be good. The stdlib uses `any` nowadays: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/encoding/json/encode.go;l=158

andrewphelpsj reviewed (approved):
Changes look good, thank you!

pedronis commented on aspects/aspects.go:
```diff
@@ -378,7 +379,7 @@ func (a *Aspect) Set(databag DataBag, request string, value interface{}) error {
 			nestedValue, ok = mapVal[part]
 			if !ok {
-				return badRequestErrorFrom(a, "set", request, `cannot key value with unmatched %q key`, part)
+				return badRequestErrorFrom(a, "set", request, `cannot find nested value with unmatched %q key`, part)
```
wouldn't "key %q" be a more consistent with other errors and slightly clearer?

MiguelPires replied:
Good point, will change that

pedronis reviewed (approved):
thank you, small comment about the error message

ernestl commented:
Failures does not relate to code changes. Force merging.
